Editorial: Model the bindings of a declarative Env Rec as a List of Records

jmdyck opened the pull request:
<strike>Making this a draft PR because I'm not sure about module Env Recs.</strike>

jmdyck committed:
Editorial: Use Declarative Env Rec's HasBinding method

... in other Declarative Env Rec methods.

jmdyck committed:
Editorial: Use Declarative Env Rec's HasBinding method

... in other Declarative Env Rec methods.

jmdyck commented:
Some alternatives for DeclarativeBindingRecord:

(1)
Replace `[[IsInitialized]]: Boolean,  [[BoundValue]]: ECMAScript language value`
with `[[BoundValue]]: ~uninitialized~ or ECMAScript language value`

(2)
To avoid the weirdness of the `[[BoundValue]]` field in indirect bindings...

(2a)
Instead of defining the sub-type ImportDeclarativeBindingRecord, allow the `[[BoundValue]]` field of a DeclarativeBindingRecord to be a record with just the `[[TargetModuleRec]]` and `[[TargetName]]` fields.

(2b)
Or you could have an abstract base type that is specialized into *either* having a `[[BoundValue]]` field *or* the `[[TargetModuleRec]]` and `[[TargetName]]` fields. (E.g. DeclarativeBindingRecord specializes to DirectDeclarativeBindingRecord and IndirectDeclarativeBindingRecord.) This is probably nicest from a theoretical standpoint, but 2a is basically equivalent without having to set up another type hierarchy.

jmdyck commented:
(forced-pushed after rebase to master)

jmdyck commented:
(force-pushed to resolve merge conflicts)

jmdyck commented:
Okay, this PR is ready for review.

Issue #2639 might affect SetMutableBinding, but probably not much.

michaelficarra commented on spec.html:
```diff
@@ -9989,6 +9989,72 @@ <h1>The Environment Record Type Hierarchy</h1>
+          </table>
+        </emu-table>
+
+        <p>A <dfn>DeclarativeBinding</dfn> is either a SimpleDeclarativeBinding or an ImportDeclarativeBinding. SimpleDeclarativeBindings can appear in any Declarative Environment Record, but ImportDeclarativeBindings can only appear in a Module Environment Record.</p>
```
```suggestion
        <p>A <dfn variants="DeclarativeBindings">DeclarativeBinding</dfn> is either a SimpleDeclarativeBinding or an ImportDeclarativeBinding. SimpleDeclarativeBindings can appear in any Declarative Environment Record, but ImportDeclarativeBindings can only appear in a Module Environment Record.</p>
```

Same for `SimpleDeclarativeBinding` and `ImportDeclarativeBinding`.

jmdyck replied:
Done.

michaelficarra commented on spec.html:
```diff
@@ -9989,6 +9989,72 @@ <h1>The Environment Record Type Hierarchy</h1>
+              <th>Meaning</th>
+            </tr>
+            <tr>
+              <td>[[BoundName]]</td>
```
This field needs to be moved to the description of `DeclarativeBinding` since we refer to it on values that are known only to be a `DeclarativeBinding`.

jmdyck replied:
Are you saying it needs to be that way in order for the spec to be "correct" in some sense? Because it sounds like you're talking about statically typing the pseudocode, and you have in mind a specific rule for how that should work.

As a counter-example, consider `ComposeWriteEventBytes`, which has `_W_.[[ByteIndex]]`, where `_W_` is either a WriteSharedMemory event or a ReadModifyWriteSharedMemory event, each of which separately declares a `[[ByteIndex]]` field. There are several other such examples.

michaelficarra commented on spec.html:
```diff
@@ -10005,7 +10071,7 @@ <h1>
           </dl>
           <emu-alg>
-            1. If _envRec_ has a binding for _N_, return *true*.
+            1. If _envRec_.[[Bindings]] contains a DeclarativeBinding whose [[BoundName]] field equals _N_, return *true*.
```
```suggestion
            1. If _envRec_.[[Bindings]] contains a DeclarativeBinding whose [[BoundName]] field is _N_, return *true*.
```

jmdyck replied:
Done.

michaelficarra commented on spec.html:
```diff
@@ -10067,9 +10135,11 @@ <h1>
-            1. Set the bound value for _N_ in _envRec_ to _V_.
-            1. <emu-not-ref>Record</emu-not-ref> that the binding for _N_ in _envRec_ has been initialized.
+            1. Assert: _envRec_.HasBinding(_N_) is *true*.
+            1. Let _binding_ be the DeclarativeBinding in _envRec_.[[Bindings]] whose [[BoundName]] field equals _N_.
```
```suggestion
            1. Let _binding_ be the DeclarativeBinding in _envRec_.[[Bindings]] whose [[BoundName]] field is _N_.
```

michaelficarra commented on spec.html:
```diff
@@ -10067,9 +10135,11 @@ <h1>
-            1. <emu-not-ref>Record</emu-not-ref> that the binding for _N_ in _envRec_ has been initialized.
+            1. Assert: _envRec_.HasBinding(_N_) is *true*.
+            1. Let _binding_ be the DeclarativeBinding in _envRec_.[[Bindings]] whose [[BoundName]] field equals _N_.
+            1. Assert: _binding_ is a SimpleDeclarativeBinding.
```
Can't we just say `SimpleDeclarativeBinding` in place of `DeclarativeBinding` in the above step then?

jmdyck replied:
I don't think the spec does that.

That is, in a phrase like:
> the [thing] in [some list] that satisfies [some constraint]

I don't think the `[thing]` phrase is ever restrictive. Rather, it's a descriptor that's true of every element of `[some list]`. All the restriction happens in `[some constraint]`.

michaelficarra commented on spec.html:
```diff
@@ -10090,15 +10160,18 @@ <h1>
               1. Return ~unused~.
-            1. If the binding for _N_ in _envRec_ is a strict binding, set _S_ to *true*.
-            1. If the binding for _N_ in _envRec_ has not yet been initialized, then
+            1. Let _binding_ be the DeclarativeBinding in _envRec_.[[Bindings]] whose [[BoundName]] field equals _N_.
```
```suggestion
            1. Let _binding_ be the DeclarativeBinding in _envRec_.[[Bindings]] whose [[BoundName]] field is _N_.
```

michaelficarra commented on spec.html:
```diff
@@ -10126,9 +10199,11 @@ <h1>
-            1. If the binding for _N_ in _envRec_ is an uninitialized binding, throw a *ReferenceError* exception.
-            1. Return the value currently bound to _N_ in _envRec_.
+            1. Assert: _envRec_.HasBinding(_N_) is *true*.
+            1. Let _binding_ be the DeclarativeBinding in _envRec_.[[Bindings]] whose [[BoundName]] field equals _N_.
```
```suggestion
            1. Let _binding_ be the DeclarativeBinding in _envRec_.[[Bindings]] whose [[BoundName]] field is _N_.
```

jmdyck committed:
fixup: add 'variants' attributes; change "equals" to "is"

jmdyck committed:
Editorial: Treat a binding as an 'alias-able' thing

... that has a distinct existence from its env rec.

jmdyck committed:
Editorial: Treat a binding as an 'alias-able' thing

... that has a distinct existence from its env rec.

jmdyck committed:
Editorial: Declarative Env Rec has field [[Bindings]]

... that is a List of bindings
(without being specific yet about what a binding is).

jmdyck committed:
Editorial: Declarative Env Rec has field [[Bindings]]

... that is a List of bindings
(without being specific yet about what a binding is).

jmdyck committed:
Editorial: Model a declarative binding as a Record

jmdyck committed:
Editorial: Model a declarative binding as a Record

linusg commented on spec.html:
```diff
@@ -10360,6 +10300,73 @@ <h1>The Environment Record Type Hierarchy</h1>
+            <tr>
+              <td>[[BoundValue]]</td>
+              <td>an ECMAScript language value or ~uninitialized~</td>
+              <td>the value that the name is bound to.</td>
```
Please capitalize the sentences in the third column

jmdyck replied:
They're not sentences, but yeah the spec styles them as if they were, so done.

linusg commented on spec.html:
```diff
@@ -11435,9 +11478,10 @@ <h1>
           <emu-alg>
-            1. Assert: _envRec_ does not already have a binding for _N_.
+            1. Assert: _envRec_.HasBinding(_N_) is *false*.
             1. Assert: When _M_.[[Environment]] is instantiated, it will have a direct binding for _N2_.
```
This assert should probably be rewritten to use `SimpleDeclarativeBinding` instead of `direct binding`

jmdyck replied:
I dunno, looking at occurrences of "direct binding", I'm not sure if that's what it means.

linusg replied:
This is only called in InitializeEnvironment with the [[Module]] and [[BindingName]] of a non-namespace ResolvedBinding, and since ResolveExport is called recursively until it finds the module containing [[BindingName]] in its [[LocalExportEntries]] it must be a SimpleDeclarativeBinding.

Additionally Module Environment's GetBindingValue would benefit from an assertion in its ImportDeclarativeBinding branch to ensure that an ImportDeclarativeBinding is always flattened and never points to another ImportDeclarativeBinding.

jmdyck replied:
> This is only called in InitializeEnvironment with the [[Module]] and [[BindingName]] of a non-namespace ResolvedBinding, and since ResolveExport is called recursively until it finds the module containing [[BindingName]] in its [[LocalExportEntries]] it must be a SimpleDeclarativeBinding.

Okay, I've changed the main clause to `its [[Bindings]] will contain a SimpleDeclarativeBinding whose [[BoundName]] field is _targetName_`.

> Additionally Module Environment's GetBindingValue would benefit from an assertion in its ImportDeclarativeBinding branch to ensure that an ImportDeclarativeBinding is always flattened and never points to another ImportDeclarativeBinding.

That's a little tricky, as the point at which that binding is currently 'exposed' is inside the invocation of `_targetEnv_.GetBindingValue`.

I suppose, before that invocation, we could say `Assert: ! _targetEnv_.HasBinding(_targetName_) is *true* and the DeclarativeBinding in _targetEnv_.[[Bindings]] whose [[BoundName]] field is _targetName_ is a SimpleDeclarativeBinding`, but that's a lot.

Instead, could we "inline" the invocation of `_targetEnv_.GetBindingValue` and assert away its ImportDeclarativeBinding branch?

linusg replied:
> Instead, could we "inline" the invocation of `_targetEnv_.GetBindingValue` and assert away its ImportDeclarativeBinding branch?

I like that suggestion, probably easier to grasp than the current recursion.

jmdyck replied:
Done.

I think I would then merge the last three steps of the ImportDeclarativeBinding branch into the last three steps of the algorithm by changing `Let _targetBinding_ be` to `Set _binding_ to`. But it's sort of a matter of taste.

linusg commented on spec.html:
```diff
@@ -10395,8 +10402,9 @@ <h1>
           <emu-alg>
-            1. Assert: _envRec_ does not already have a binding for _N_.
-            1. Create a mutable binding in _envRec_ for _N_ and record that it is uninitialized. If _D_ is *true*, record that the newly created binding may be deleted by a subsequent DeleteBinding call.
+            1. Assert: _envRec_.HasBinding(_N_) is *false*.
```
missing `!`

jmdyck replied:
Thanks, I should have seen that (and the others).

linusg commented on spec.html:
```diff
@@ -10416,8 +10424,9 @@ <h1>
           <emu-alg>
-            1. Assert: _envRec_ does not already have a binding for _N_.
-            1. Create an immutable binding in _envRec_ for _N_ and record that it is uninitialized. If _S_ is *true*, record that the newly created binding is a strict binding.
+            1. Assert: _envRec_.HasBinding(_N_) is *false*.
```
Missing `!`

linusg commented on spec.html:
```diff
@@ -10437,9 +10446,11 @@ <h1>
-            1. Assert: _envRec_ must have an uninitialized binding for _N_.
-            1. Set the bound value for _N_ in _envRec_ to _V_.
-            1. <emu-not-ref>Record</emu-not-ref> that the binding for _N_ in _envRec_ has been initialized.
+            1. Assert: _envRec_.HasBinding(_N_) is *true*.
```
Missing `!`

linusg commented on spec.html:
```diff
@@ -10460,15 +10471,18 @@ <h1>
           </dl>
           <emu-alg>
-            1. [id="step-setmutablebinding-missing-binding"] If _envRec_ does not have a binding for _N_, then
+            1. [id="step-setmutablebinding-missing-binding"] If _envRec_.HasBinding(_N_) is *false*, then
```
Missing `!`

linusg commented on spec.html:
```diff
@@ -10496,9 +10510,11 @@ <h1>
-            1. Assert: _envRec_ has a binding for _N_.
-            1. If the binding for _N_ in _envRec_ is an uninitialized binding, throw a *ReferenceError* exception.
-            1. Return the value currently bound to _N_ in _envRec_.
+            1. Assert: _envRec_.HasBinding(_N_) is *true*.
```
Missing `!`

linusg commented on spec.html:
```diff
@@ -10516,9 +10532,11 @@ <h1>
-            1. Assert: _envRec_ has a binding for _N_.
-            1. If the binding for _N_ in _envRec_ cannot be deleted, return *false*.
-            1. Remove the binding for _N_ from _envRec_.
+            1. Assert: _envRec_.HasBinding(_N_) is *true*.
```
Missing `!`

linusg commented on spec.html:
```diff
@@ -11374,14 +11414,17 @@ <h1>
-            1. Assert: _envRec_ has a binding for _N_.
-            1. If the binding for _N_ is an indirect binding, then
-              1. Let _M_ and _N2_ be the indirection values provided when this binding for _N_ was created.
+            1. Assert: _envRec_.HasBinding(_N_) is *true*.
```
Missing `!`

linusg commented on spec.html:
```diff
@@ -11435,9 +11478,10 @@ <h1>
           </dl>
           <emu-alg>
-            1. Assert: _envRec_ does not already have a binding for _N_.
+            1. Assert: _envRec_.HasBinding(_N_) is *false*.
```
Missing `!`

bakkot commented on spec.html:
```diff
@@ -10343,6 +10343,73 @@ <h1>The Environment Record Type Hierarchy</h1>
+                a List of DeclarativeBindings
+              </td>
+              <td>
+                Satisfies the invariant that no two records in [[Bindings]] have the same [[BoundName]].
```
This isn't really a "meaning". It is a property this has but not what it means.

Maybe:

```suggestion
                The bindings in this Environment Record. Satisfies the invariant that no two records in [[Bindings]] have the same [[BoundName]].
```

jmdyck replied:
Done.

bakkot commented on spec.html:
```diff
@@ -10343,6 +10343,73 @@ <h1>The Environment Record Type Hierarchy</h1>
+          </table>
+        </emu-table>
+
+        <p>A <dfn variants="DeclarativeBindings">DeclarativeBinding</dfn> is either a SimpleDeclarativeBinding or an ImportDeclarativeBinding. SimpleDeclarativeBindings can appear in any Declarative Environment Record, but ImportDeclarativeBindings can only appear in a Module Environment Record.</p>
```
I was surprised to read this and then find only the definition of SimpleDeclarativeBinding, so perhaps:

```suggestion
        <p>A <dfn variants="DeclarativeBindings">DeclarativeBinding</dfn> is either a SimpleDeclarativeBinding or an ImportDeclarativeBinding. SimpleDeclarativeBindings can appear in any Declarative Environment Record, but ImportDeclarativeBindings can only appear in a Module Environment Record. The definition of ImportDeclarativeBinding is in <emu-xref href="#sec-module-environment-records"></emu-xref>.</p>
```

jmdyck replied:
Done.

bakkot commented on spec.html:
```diff
@@ -10443,15 +10514,18 @@ <h1>
               1. Throw a *ReferenceError* exception.
-            1. Else if the binding for _name_ in _envRec_ is a mutable binding, then
+            1. Else if _binding_.[[IsMutable]] is *true*, then
               1. Change its bound value to _value_.
```
```suggestion
              1. Set _binding_.[[BoundValue]] to _value_.
```

jmdyck replied:
Huh, odd I missed that one.

bakkot commented on spec.html:
```diff
@@ -10343,6 +10343,73 @@ <h1>The Environment Record Type Hierarchy</h1>
+          </table>
+        </emu-table>
+
+        <p>The behaviour of the concrete specification methods for Declarative Environment Records is defined by the following algorithms.</p>
```
```suggestion
        <p>The default behaviour of the concrete specification methods for Declarative Environment Records is defined by the following algorithms; Function Environment Records and Module Environment Records are specialized kinds of Declarative Environment Records which provide alternative specifications for some of these methods.</p>
```

jmdyck replied:
Done.

bakkot commented on spec.html:
```diff
@@ -52150,7 +52262,7 @@ <h1>Races</h1>
       1. If _eventA_ and _eventB_ are not the same Shared Data Block event, then
         1. If it is not the case that both _eventA_ happens-before _eventB_ in _execution_ and _eventB_ happens-before _eventA_ in _execution_, then
-          1. If _eventA_ and _eventB_ are both WriteSharedMemory or ReadModifyWriteSharedMemory events and _eventA_ and _eventB_ do not have disjoint memory ranges, then
+          1. If _eventA_ is either a WriteSharedMemory or ReadModifyWriteSharedMemory event, _eventB_ is either a WriteSharedMemory or ReadModifyWriteSharedMemory event, and _eventA_ and _eventB_ do not have disjoint memory ranges, then
```
Stray?

jmdyck replied:
A rebase transient, fixed by PR #3839.

bakkot reviewed (approved):
This is a huge improvement. Some minor tweaks but otherwise looks good.

jmdyck committed:
fixup: Capitalize the "Meaning" of fields

jmdyck committed:
fixup: Insert `!` before HasBinding invocations

jmdyck committed:
fixup: Better "Meaning" for `[[Bindings]]`

jmdyck committed:
fixup: Add forward-reference for ImportDeclarativeBinding

jmdyck committed:
fixup: Convert "Change its bound value"

jmdyck committed:
fixup: Mention that Function ERs and Module ERs override certain concrete methods

jmdyck committed:
Editorial: Mention that Function ERs and Module ERs override certain concrete methods

jmdyck committed:
fixup: rephrase "it will have a direct binding for" in CreateImportBinding assert

jmdyck committed:
fixup: Unfold `_targetEnv_.GetBindingValue(...)` in GetBindingValue for Module Environment Records
